homeassistant/components/google_mail/application_credentials.py

```diff
@@ -25,3 +33,11 @@ async def async_get_description_placeholders(hass: HomeAssistant) -> dict[str, s
         "more_info_url": "https://www.home-assistant.io/integrations/google_mail/",
         "oauth_creds_url": "https://console.cloud.google.com/apis/credentials",
     }
+
+
+async def get_oauth_service(data: dict[str, Any]) -> Resource:
+    """Get valid service with latest access token."""
+    session: OAuth2Session = data[DATA_SESSION]
+    await session.async_ensure_token_valid()
+    credentials = Credentials(data[CONF_TOKEN][CONF_ACCESS_TOKEN])
```

Comment:
```suggestion
    credentials = Credentials(session.token[CONF_ACCESS_TOKEN])
```

Reply:
FWIW i think these scaffold give pretty good patterns for doing this:
https://github.com/home-assistant/core/blob/6d4489e82207916a811e936be00246c5cbc10545/script/scaffold/templates/config_flow_oauth2/integration/api.py#L49

... so you don't have to pass around data everywhere, you can just pass around an `Auth` client library that gives you back a `Resource`.